Allow MSP on VCP, stub out another message handler

mlyle opened the pull request:
Open for transparency

mlyle committed:
MSPBridge: reformat command list, add MSP_SET_4WAY_IF

mlyle committed:
UAVOMSPBridge: Remove "slow" uavtalk detection

GCS will no longer try to speak 57.6k by default (for quite some time).
Also no more Naze32, which was the main reason for this.

mlyle committed:
pios_hal: MSP on USB CDC VCP TLAs

mlyle committed:
uavomspbridge: stub out the ESC passthrough command

mlyle committed:
uavomspbridge: USB does not like that long of timeout

mlyle committed:
targets: correct UAVO missing parent for VCP

mlyle commented:
This works, and is a good first step IMO.

tracernz commented on flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c:
```diff
@@ -132,13 +133,22 @@ typedef enum {
-	MSP_MAYBE_UAVTALK_SLOW6
 } msp_state;
 
+typedef enum __attribute__ ((__packed__)) {
```
Not a fan of typedefing this. :/

tracernz commented on flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c:
```diff
@@ -662,27 +695,6 @@ static bool msp_receive_byte(struct msp_bridge *m, uint8_t b)
-			PIOS_COM_TELEM_RF = m->com;
-			return false;
-		}
-		break;
```
Fair enough I guess.

tracernz commented on flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c:
```diff
@@ -754,7 +766,9 @@ static void uavoMSPBridgeTask(void *parameters)
 		uint8_t b = 0;
-		uint16_t count = PIOS_COM_ReceiveBuffer(msp->com, &b, 1, PIOS_QUEUE_TIMEOUT_MAX);
+
+		uint16_t count = PIOS_COM_ReceiveBuffer(msp->com, &b, 1, 3000);
```
Any reason for 3000? It won't do anything other than unblock this task so it can immediately block again AFAICT?

mlyle replied:
There's some kind of issue with PIOS_QUEUE_TIMEOUT_MAX (which isn't even a PIOS_COM value) and USB/VCP where it hangs forever even if there's bytes to read.  3000 is chosen to be pretty long / not burden us too much.

tracernz replied:
Heh, okay. Maybe mark a TODO to look at why it behaves like that?

tracernz commented on flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c:
```diff
@@ -487,7 +498,7 @@ static void msp_send_channels(struct msp_bridge *m)
 		.channels = {
 			msp_scale_rc(manualState.Roll),
-			msp_scale_rc(manualState.Pitch * -1), // TL pitch is backwards
+			msp_scale_rc(manualState.Pitch * -1), // MW pitch is backwards
```
:D

yamahabest commented on shared/uavobjectdefinition/hwplayuavosd.xml:
```diff
@@ -15,7 +15,7 @@
 			<description/>
 		</field>
-		<field name="USB_VCPPort" units="function" type="enum" elements="1" options="USBTelemetry,ComBridge,DebugConsole,Disabled" defaultvalue="Disabled">
+		<field name="USB_VCPPort" units="function" type="enum" elements="1" options="USBTelemetry,ComBridge,DebugConsole,Disabled" parent="HwShared.USB_VCPPort" defaultvalue="Disabled">
```
You have removed the "options" attribute for the others, when you added the "parent" attribute.
Can't the "options" attribute also be removed here? The "parent" attribute already dictates these options, right?

mlyle replied:
It's a subset here not including MSP.

tracernz replied:
I almost included UAVO-to-MSP on this board. At that point it could act as an MSP relay, but I couldn't really see a use-case for it.

mlyle replied:
Could hook a mwosd to playuavosd to a seppuku and have an OSD on OSD on OSD ;)

yamahabest commented on shared/uavobjectdefinition/hwtaulink.xml:
```diff
@@ -23,7 +23,7 @@
 			<description/>
 		</field>
-		<field name="VCPPort" units="" type="enum" elements="1" parent="HwShared.USB_VCPPort" defaultvalue="Disabled">
+		<field name="VCPPort" units="" type="enum" elements="1" parent="HwShared.USB_VCPPort" options="USBTelemetry,ComBridge,DebugConsole,Disabled" defaultvalue="Disabled">
```
Same as hwplayuavosd.xml comment.

mlyle replied:
Also a subset not including MSP.

yamahabest commented on shared/uavobjectdefinition/hwaq32.xml:
```diff
@@ -123,7 +123,7 @@
 			<description/>
 		</field>
-		<field name="USB_VCPPort" units="function" type="enum" elements="1" options="USBTelemetry,ComBridge,DebugConsole,Disabled" defaultvalue="Disabled">
+		<field name="USB_VCPPort" units="function" type="enum" elements="1" parent="HwShared.USB_VCPPort" defaultvalue="Disabled">
```
When the "parent" sets a defaultvalue, does the "child" still need to set a defaultvalue?
Or should the defaultvalue for these 4 boards not be "MSP"?

mlyle replied:
It may be possible with further discussion to pull the "Disabled" off these but I didn't feel like advocating for it with this PR (it will consume RAM but be nice in various ways).

tracernz commented on flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c:
```diff
@@ -132,13 +133,22 @@ typedef enum {
+} msp_esc_protocol;
+
+struct msp_cmddata_escserial {
+	msp_esc_protocol protocol;
```
I.e. it actively hides what protocol is here, and for any type hints in IDEs etc.

mlyle replied:
I don't understand..

Is your objection the typedef and you'd prefer enum msp_esc_protocol here?  Or do you want me to use uint8_t like a lot of our other stuff does  Or is it variable name?

Things are already fairly (descriptively) buried in `m->cmd_data.escserial.protocol` --  that's why I rejected making it `m->cmd_data.cmd_escserial.esc_protocol`

tracernz replied:
> Is your objection the typedef and you'd prefer enum msp_esc_protocol here?

Yes, exactly.

mlyle replied:
OK, I will change.  Note it's likely to go back, though, if we take BetaFlight's implementation which does:

```
typedef enum {
    PROTOCOL_SIMONK = 0,
    PROTOCOL_BLHELI = 1,
    PROTOCOL_KISS = 2,
    PROTOCOL_KISSALL = 3,
    PROTOCOL_CASTLE = 4,
    PROTOCOL_COUNT
} escProtocol_e;
```

it will go back.

tracernz commented on shared/uavobjectdefinition/hwshared.xml:
```diff
@@ -67,7 +67,7 @@
 		<field name="MagOrientation" units="function" type="enum" elements="1" options="Top0degCW,Top90degCW,Top180degCW,Top270degCW,Bottom0degCW,Bottom90degCW,Bottom180degCW,Bottom270degCW" defaultvalue="Top0degCW"/>
 		<field name="USB_HIDPort" units="function" type="enum" elements="1" options="USBTelemetry,Disabled" defaultvalue="USBTelemetry"/>
-		<field name="USB_VCPPort" units="function" type="enum" elements="1" options="USBTelemetry,ComBridge,DebugConsole,Disabled" defaultvalue="Disabled"/>
+		<field name="USB_VCPPort" units="function" type="enum" elements="1" options="USBTelemetry,ComBridge,DebugConsole,MSP,Disabled" defaultvalue="MSP"/>
```
Is this default change intentional? I expect/hope not.

mlyle replied:
Right now all targets override it.

It's kinda neat when default is MSP, because of the fallback-to-telemetry behavior.  It allows you to use VCP telemetry without conflicting over USB telem ids / that legacy infrastructure.

tracernz replied:
(Ignoring boards overriding) This will cause the MSP module to always run. That's a marked change from our current philosophy that warrants further discussion IMO.

mlyle replied:
Yes.  It's a definite RAM cost.

Advantage would be that passthrough would work out of box, and VCP telemetry would work out of box, too (could be a step along the way to transitioning away from HID telemetry).

In any case, it's not functional because every target currently overrides.

mlyle committed:
msp: support more messages

mlyle commented:
Rebased, added a few more messages.

With this, blheliconfigurator is able to talk to the board:


SERIAL: Connection opened with ID: 1, Baud: 115200  
Serial port <span style="color: green">successfully</span> opened with ID: 1  
MultiWii API version <span style="color: green">received</span> - <strong>1.31.0</strong>  
Flight controller info, identifier: <strong>sepp</strong>, version: <strong>1.0.0</strong>  
Running firmware released on: <strong></strong>  
Board: <strong>sepp</strong>, version: <strong>0</strong>  
Unique device ID <span style="color: green">received</span> -   <strong>0x1b003a3435511837303638</strong>

mlyle commented:
Annoyance: VCP serial ports do not work for several seconds after connection (9-10 seconds).

mlyle committed:
uavomspbridge: send FC id as firmware

Because apparently this is what it means.

mlyle commented:
Changing FC ID to "BTFL" makes Betaflight Configurator connect:

```
2017-03-29 @ 12:26:08 -- Serial port successfully opened with ID: 9
2017-03-29 @ 12:26:08 -- MultiWii API version received - 1.31.0
2017-03-29 @ 12:26:08 -- Flight controller info, identifier: BTFL, version: 1.0.0
2017-03-29 @ 12:26:08 -- Running firmware released on:
2017-03-29 @ 12:26:08 -- Board: sepp, version: 0
2017-03-29 @ 12:26:08 -- Unique device ID received - 0x1b003a3435511837303638
```

mlyle commented:
On my WIndows machine, blhelisuite fails.  Debug log in blheli shows it sending sensible things and getting no response.  Typing in those sensible things in putty-to-serial-port gets response.  Will require further investigation.

mlyle commented:
@tracernz your comment about PIOS_QUEUE_TIMEOUT_MAX has disappeared, but I opened #1679 to cover it and other issues.

tracernz commented:
Thanks.

mlyle committed:
msp: add msgs so blhelisuite gets to tryng 4wif

mlyle committed:
uavomspbridge: send error on unknown message

also put the enum back

mlyle commented:
With the last 2 changes, one gets:

![blhelisuitelogwindow_170329_1](https://cloud.githubusercontent.com/assets/90903/24479901/6d2a04b2-1496-11e7-861c-1e15a47d2b06.png)

tracernz commented:
That's the state I got to before with the pure 4W-IF stuff. Now the hard bits! 😜

mlyle commented:
@tracernz I actually think with your stuff I'd get a couple more steps (as 4wifing has just started at the bottom of the log (and is ignored)

mlyle committed:
hwshared: for now set default VCP "Disabled"

mlyle commented:
Set the default value of VCP to disabled for now, which should make this less controversial.  Mergey-merge pls?
